src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2022 The Raksha Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+#define SRC_IR_AUTH_LOGIC_SOUFFLE_EMITTER_H_
+
+#include <unordered_set>
+
+#include "src/ir/auth_logic/ast.h"
+
+namespace raksha::ir::auth_logic {
+
+class SouffleEmitter {
+  public:
+    std::string EmitProgram(DLIRProgram program) {
+      SouffleEmitter emitter;
+      std::string body = emitter.EmitProgramBody(program);
+      std::string outputs = emitter.EmitOutputs(program);
+      std::string decls = emitter.EmitDeclarations();
+      return std::move(body) 
+        + "\n" 
+        + std::move(outputs) 
+        + "\n" 
+        + std::move(decls);
```

Comment:
Use `absl::StrCat` here instead of multiple concatenations. Also, you can avoid the need for `std::move` if you never give the values names in the first place; you could just not make the `body`, `outputs`, and `decls` locals and inline them directly into the call to `StrCat`.

Reply:
fixed